### PR TITLE
Refactor guid, message, queue

### PR DESCRIPTION
**Pull request by Dieterbe**
what do we think about this?

note: currently causes `go test` to fail.  haven't figured it out yet, wanted to check first if the general idea looks good.

### CODE REVIEW

File: nsqd/backend_queue.go [hunk 1]
```diff
@@ -9,4 +11,5 @@ type BackendQueue interface {
 	Delete() error
 	Depth() int64
 	Empty() error
+	WriteMsg(buf *bytes.Buffer, msg *Message) error
```
Comment (mreiferson) [inline nsqd/backend_queue.go]: It was intentional for `DiskQueue` to not have first class knowledge of `Message` - that's why we had the shim in between the two.

File: nsqd/guid/guid.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package nsqd
+package guid
```
Comment (mreiferson) [inline nsqd/guid/guid.go]: this whole package would be going away in #625, so I wouldn't spin your wheels on it

**Comment by mreiferson** [comment]
@Dieterbe I appreciate the effort to clean things up.  As per my two comments above, I don't think we would land this.

If you want to get your feet wet with the codebase, might I suggest some of the [`contributor-friendly` issues](https://github.com/bitly/nsq/issues?q=is%3Aopen+is%3Aissue+label%3Acontributor-friendly)?

**Comment by Dieterbe** [comment]
i see. totally get DiskQueue should not deal with message concepts, makes total sense.
(after all i've also yanked out the diskqueue for user in other projects)

the main thing i wanted to achieve is make Guid public so that i could embed nsqd into my app publish messages through the go api instead of tcp, the only thing seemingly needed for that is the Guid stuff.

but i read your comment on the mailing list that you're not a big fan approach, so maybe i ishould just deal with it and use the network api. (through the go-nsq library) though i still think direct calls are a lot more robust.

**Comment by mreiferson** [comment]
> but i read your comment on the mailing list that you're not a big fan approach, so maybe i ishould just deal with it and use the network api. (through the go-nsq library) though i still think direct calls are a lot more robust.

It's really just that we don't want to have the burden of publicly supporting the in-process API.  Besides, I think it would result in a deployment that is too tightly coupled.  NOTE: You can still generate your own message IDs, without any of the changes in this PR.

Going to close this.

Thanks again!

**Comment by Dieterbe** [comment]
thank you Matt, for the nice work :)

**Pull request closed without merging**